src/renderer/dx/CustomTextRenderer.cpp

```diff
@@ -220,6 +222,144 @@ using namespace Microsoft::Console::Render;
+try
+{
+    if (!drawingContext.cursorInfo.has_value())
+    {
+        return S_FALSE;
+    }
+
+    const auto& options = drawingContext.cursorInfo.value();
+
+    // if the cursor is off, do nothing - it should not be visible.
+    if (!options.isOn)
+    {
+        return S_FALSE;
+    }
+
+    // TODO GH#6338: Add support for `"cursorTextColor": null` for letting the
+    // cursor draw on top again.
+    if (!firstPass)
+    {
+        return S_FALSE;
+    }
+
+    const til::size glyphSize{ til::math::flooring,
+                               drawingContext.cellSize.width,
+                               drawingContext.cellSize.height };
+
+    // Create rectangular block representing where the cursor can fill.
+    D2D1_RECT_F rect = til::rectangle{ til::point{ options.coordCursor } }.scale_up(glyphSize);
+
+    // If we're double-width, make it one extra glyph wider
+    if (options.fIsDoubleWidth)
+    {
+        rect.right += glyphSize.width();
+    }
+
+    // If the cursor isn't within the bounds of this current run of text, do nothing.
+    if (rect.top > textRunBounds.bottom ||
+        rect.bottom <= textRunBounds.top ||
+        rect.left > textRunBounds.right ||
+        rect.right <= textRunBounds.left)
+    {
+        return S_FALSE;
+    }
+
+    CursorPaintType paintType = CursorPaintType::Fill;
+
+    switch (options.cursorType)
+    {
+    case CursorType::Legacy:
+    {
+        // Enforce min/max cursor height
+        ULONG ulHeight = std::clamp(options.ulCursorHeightPercent, MinCursorHeightPercent, MaxCursorHeightPercent);
+        ulHeight = (glyphSize.height<ULONG>() * ulHeight) / 100;
+        rect.top = rect.bottom - ulHeight;
+        break;
+    }
+    case CursorType::VerticalBar:
+    {
+        // It can't be wider than one cell or we'll have problems in invalidation, so restrict here.
+        // It's either the left + the proposed width from the ease of access setting, or
+        // it's the right edge of the block cursor as a maximum.
+        rect.right = std::min(rect.right, rect.left + options.cursorPixelWidth);
+        break;
+    }
+    case CursorType::Underscore:
+    {
+        rect.top = rect.bottom - 1;
+        break;
+    }
+    case CursorType::EmptyBox:
+    {
+        paintType = CursorPaintType::Outline;
+        break;
+    }
+    case CursorType::FullBox:
+    {
+        break;
+    }
+    default:
+        return E_NOTIMPL;
+    }
+
+    Microsoft::WRL::ComPtr<ID2D1SolidColorBrush> brush;
+
+    if (options.fUseColor)
+    {
+        // Make sure to make the cursor opaque
+        RETURN_IF_FAILED(d2dContext->CreateSolidColorBrush(til::color{ OPACITY_OPAQUE | options.cursorColor },
+                                                           &brush));
+    }
+
+    switch (paintType)
+    {
+    case CursorPaintType::Fill:
+    {
+        d2dContext->FillRectangle(rect, brush.Get());
+        break;
+    }
+    case CursorPaintType::Outline:

```

Comment:
i don't love that text will strike on top of the line cursors... consider what happens if your cursor is over a U+2588 "full block".

I feel like only the filling cursors should have text struck on top of them for now.

Reply:
Yea okay, that's not hard

Reply:
I think striking over or under a full block with the a cursor explicitly defined as the same color and any shape is always going to be a problem... unless we reintroduce invert.

I'd sort of want to see the visual comparison of what happens here for all the cursors either way. I'm having a hard time visualizing it in my head.